apps/arweave/src/ar_tx_poller.erl

```diff
@@ -0,0 +1,160 @@
+%% never receives transactions by push), we wait this long starting from
+%% the moment we join the network only once and then keep polling
+%% for transactions more frequently.
+-ifdef(AR_TEST).
+-define(CHECK_INTERVAL_MS, 5_000).
+-else.
+-define(CHECK_INTERVAL_MS, 30_000).
+-endif.
+
+%% Poll interval in milliseconds - how long we wait before downloading a new
+%% transaction or polling the mempools for new transactions.
+-ifdef(AR_TEST).
+-define(POLL_INTERVAL_MS, 500).
+-else.
+-define(POLL_INTERVAL_MS, 200).
+-endif.
+
+%%% Public API.
+
+start_link() ->
+	gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+%%% Gen server callbacks.
+
+init([]) ->
+    [ok, ok] = ar_events:subscribe([tx, node_state]),
+	{ok, #state{}}.
+
+handle_call(Request, From, State) ->
+	?LOG_WARNING("Unexpected call: ~p from ~p", [Request, From]),
+	{reply, ignored, State}.
+
+handle_cast(check_for_received_txs, State) ->
+	%% Check if there have been any transactions received in the last
+	%% ?CHECK_INTERVAL_MS milliseconds.
+	TimestampDiff = erlang:system_time(microsecond) - State#state.last_seen_tx_timestamp,
+	State3 =
+		case TimestampDiff > 0 andalso TimestampDiff > ?CHECK_INTERVAL_MS of
+			true ->
+				State2 = check_for_received_txs(State),
+				ar_util:cast_after(?POLL_INTERVAL_MS, self(), check_for_received_txs),
+				State2;
+			false ->
+				ar_util:cast_after(?CHECK_INTERVAL_MS, self(), check_for_received_txs),
+				State
+		end,
+	{noreply, State3};
+
+handle_cast(Request, State) ->
+	?LOG_WARNING("Unexpected cast: ~p", [Request]),
+	{noreply, State}.
+
+handle_info({event, node_state, {initialized, _}}, State) ->
+	%% Send a check_for_received_txs cast periodically to check for externally submitted
+	%% transactions. If there have not been any for longer than 30 seconds, request the
+	%% mempool from a peer and download the transactions.
+    {ok, Config} = application:get_env(arweave, config),
+    case lists:member(tx_poller, Config#config.disable) of
+        true ->
+            ok;
+        false ->
+            gen_server:cast(self(), check_for_received_txs)
+    end,
+    {noreply, State};
+
+handle_info({event, node_state, _}, State) ->
+	{noreply, State};
+
+handle_info({event, tx, {new, _TX, {pushed, _Peer}}}, State) ->
+	{noreply, State#state{
+		pending_txids = [],
+		last_seen_tx_timestamp = erlang:system_time(microsecond)
+	}};
+
+handle_info({event, tx, _}, State) ->
+	{noreply, State};
+
+handle_info(Info, State) ->
+	?LOG_WARNING("event: unhandled_info, info: ~p", [Info]),
+	{noreply, State}.
+
+terminate(Reason, _State) ->
+	?LOG_WARNING("Unexpected terminate: ~p", [Reason]),
+	ok.
+
+%%% Internal functions.
+
+check_for_received_txs(#state{ pending_txids = [TXID | PendingTXIDs] } = State) ->
+	case ar_mempool:is_known_tx(TXID) of
+		true ->
+			ok;
+		false ->
+			download_and_verify_tx(TXID)
+	end,
+	State#state{ pending_txids = PendingTXIDs };
+
+check_for_received_txs(#state{ pending_txids = [] } = State) ->
```

Comment:
When this clause is hit, that restarts the process, right? So the flow is like
1. If no tx has been pushed in CHECK_INTERVAL_MS, start the `check_for_received_txs` loop
2. First get the the mempool of a peer
3. Wait POLL_INTERVAL_MS then query a tx from some per
4. Wait POLL_INVERVAL_MS then query the next tx from some peer
5. repeat 3,4 until the pending list is empty, then return to 2
6. repeat 2-5 until a tx is pushed to the node

Reply:
correct